### PR TITLE
WIP comparing swagger-codegen-cli-v3 to openapi-generator-cli

### PR DESCRIPTION
**Pull request by ctreatma**
I don't have context on the original decision to adopt openapi-generator instead of swagger-codegen; the former started out as a fork of the latter, and they're both actively developed and source-available (I don't know current licensing off-hand, so I can't say if either or both are Open Source).

This is just a draft PR that we can look at to see how code generated with swagger-codegen differs from code generated with openapi-generator.

### CODE REVIEW

File: metal/v1/model_device.go [hunk 1]
```diff
@@ -1,1680 +1,71 @@
...
+	SshKeys      []Href  `json:"ssh_keys,omitempty"`
 	// The current state the instance is in.  * When an instance is initially created it will be in the `queued` state until it is picked up by the provisioner. * Once provisioning has begun on the instance it's state will move to `provisioning`. * When an instance is deleted, it will move to `deprovisioning` state until the deprovision is completed and the instance state moves to `deleted`. * If an instance fails to provision or deprovision it will move to `failed` state. * Once an instance has completed provisioning it will move to `active` state. * If an instance is currently powering off or powering on it will move to `powering_off` or `powering_on` states respectively.  * When the instance is powered off completely it will move to the `inactive` state. * When an instance is powered on completely it will move to the `active` state. * Using the reinstall action to install a new OS on the instance will cause the instance state to change to `reinstalling`. * When the reinstall action is complete the instance will move to `active` state.
-	State   *string  `json:"state,omitempty"`
+	State   string   `json:"state,omitempty"`
```
Comment (ctreatma) [inline metal/v1/model_device.go:58]: swagger-codegen _also_ doesn't handle inline enums

File: metal/v1/model_device.go [hunk 2]
```diff
@@ -1,1680 +1,71 @@
...
+	Project            *Project                `json:"project,omitempty"`
+	ProjectLite        *AllOfDeviceProjectLite `json:"project_lite,omitempty"`
+	ProvisioningEvents []Event                 `json:"provisioning_events,omitempty"`
 	// Only visible while device provisioning
-	ProvisioningPercentage *float32 `json:"provisioning_percentage,omitempty"`
+	ProvisioningPercentage float32 `json:"provisioning_percentage,omitempty"`
 	// Root password is automatically generated when server is provisioned and it is removed after 24 hours
-	RootPassword *string `json:"root_password,omitempty"`
-	ShortId      *string `json:"short_id,omitempty"`
+	RootPassword string `json:"root_password,omitempty"`
```
Comment (ctreatma) [inline metal/v1/model_device.go:50]: swagger-codegen doesn't use pointers for optional fields of built-in types; that lines up with how packngo does things

File: metal/v1/model_device.go [hunk 3]
```diff
@@ -1,1680 +1,71 @@
...
-	Userdata        *string    `json:"userdata,omitempty"`
-	Volumes         []Href     `json:"volumes,omitempty"`
+	TerminationTime time.Time `json:"termination_time,omitempty"`
+	UpdatedAt       time.Time `json:"updated_at,omitempty"`
+	User            string    `json:"user,omitempty"`
+	Userdata        string    `json:"userdata,omitempty"`
+	Volumes         []Href    `json:"volumes,omitempty"`
 	// Hostname used to connect to the instance via the SOS (Serial over SSH) out-of-band console.
-	Sos                  *string `json:"sos,omitempty"`
-	AdditionalProperties map[string]interface{}
```
Comment (ctreatma) [inline metal/v1/model_device.go:78]: `AdditionalProperties` disappeared, but that could be a configuration issue (with openapi-generator we had to tell the generator to generate spec-compliant code with AdditionalProperties support, and it's possible a similar config option is needed with swagger-codegen)

File: metal/v1/model_device.go [hunk 4]
```diff
@@ -1,1680 +1,71 @@
...
+	BondingMode         int32                 `json:"bonding_mode,omitempty"`
+	CreatedAt           time.Time             `json:"created_at,omitempty"`
+	CreatedBy           *AllOfDeviceCreatedBy `json:"created_by,omitempty"`
+	Customdata          ModelMap              `json:"customdata,omitempty"`
+	Description         string                `json:"description,omitempty"`
+	Facility            *Facility             `json:"facility,omitempty"`
+	HardwareReservation *HardwareReservation  `json:"hardware_reservation,omitempty"`
+	Hostname            string                `json:"hostname,omitempty"`
+	Href                string                `json:"href,omitempty"`
+	Id                  string                `json:"id,omitempty"`
```
Comment (ctreatma) [inline metal/v1/model_device.go:28]: Initialisms also seem to be a problem for swagger-codegen

File: metal/v1/model_device.go [hunk 5]
```diff
@@ -1,1680 +1,71 @@
...
-	ImageUrl            *string                `json:"image_url,omitempty"`
-	IpAddresses         []IPAssignment         `json:"ip_addresses,omitempty"`
-	IpxeScriptUrl       *string                `json:"ipxe_script_url,omitempty"`
-	Iqn                 *string                `json:"iqn,omitempty"`
+	AlwaysPxe           bool                  `json:"always_pxe,omitempty"`
+	BillingCycle        string                `json:"billing_cycle,omitempty"`
+	BondingMode         int32                 `json:"bonding_mode,omitempty"`
+	CreatedAt           time.Time             `json:"created_at,omitempty"`
+	CreatedBy           *AllOfDeviceCreatedBy `json:"created_by,omitempty"`
+	Customdata          ModelMap              `json:"customdata,omitempty"`
```
Comment (ctreatma) [inline metal/v1/model_device.go:22]: openapi-generator definitely gives a nicer result for this attribute and the one above it

**Comment by displague** [comment]
Prior to openapi-generator, we were using go-swagger generator. go-swagger wasn't OAS3 ready and we were moving from Swagger to OAS3.  

I don't recall how deeply swagger-codegen was considered at the time. I was aware of the split and Github liveliness may have been part of the decision.  OAS3 features between swagger-codegen and openapi-generator may have been at different maturity levels. It's possible the legacy protocol name scared me off.

Check out https://github.com/equinix-labs/metal-go/pull/31 for the diff from that exploration.  As a result, we lost #26 which the go-swagger generator offered.

**Comment by ctreatma** [comment]
Closing this; it only exists to compare the results of the generators, not as a real proposal.  The feature sets of the 2 generators seem roughly identical, and the differences seem largely stylistic.

**Pull request closed without merging**